### ISSUE TITLE
arraySplice doesn't trigger render

### ISSUE THREAD

**Issue by nikordaris** [redux-form/redux-form#3351]
The action creator ```arraySplice``` when used to replace a single item in the array doesn't triggering a render with the updated values in the ```FieldArray.fields```. The store gets correctly updated but React doesn't detect a change in the array since it is the same instance and length. I expected this action creator to account for this by cloning the array when the ```removeNum``` is 1. If this is the intended behavior perhaps an additional action creator needs to exist to update a single item in the array properly.

redux-form@7.0.3
react@15.6.1
redux@3.7.1

### Works
```javascript
const items = fields.getAll();
items.splice(index, 1, values);
change(form, field, [...items]);
```

### Doesn't Work
```javascript
arraySplice(form, field, index, 1, values);
```

**Comment by nikordaris** [comment]
This might be related to #3281. I'm noticing ```change``` works for the first edit but doesn't trigger rendering for follow on edits.

**Comment by nikordaris** [comment]
I've confirmed that the core issue here with ```arraySplice``` is unrelated to #3281 but my workaround suffers from the same issue. The workaround for my workaround is found https://github.com/erikras/redux-form/issues/3281#issuecomment-321309288

**Comment by slavab89** [comment]
Same thing happens to be when using remove&push for FieldArray. It appears there was an issue like this before https://github.com/erikras/redux-form/issues/2466 and it should have been fixed but i still experience this.

**Comment by priand** [comment]
Same here with fields.push() & fields.removeAll(): the component render() is not triggered. By the way, there is not splice() method in the fields prop.
Bottom line: I'm using a single input to edit an array of strings using a separator (,). I set the input value using join(fields.getAll()), and on onChange, I split() the input string to make it an array again. Then I replace the array in the store call calling remove() and push(). If you know a better solution, please let me know :-)
l'll try to call change() as @nikordaris suggested above

**Comment by priand** [comment]
I tried the suggestion above:
```
            let a = e.target.value.split(',')
            form.change(fields.name,a);
```
But I'm getting the same behavior. The value is updated in the store (I'm displaying the JSON) vut the FieldArray component is not rendered()

Update: I got it to work using rerenderOnEveryChange=true, either using the fields method or a call to change(). That might be the solution actually.

**Comment by slavab89** [comment]
Yeah you can do that, or use a selector for the field or `formValueSelector`. The thing is that once you do `rerenderOnEveryChange` it will impact performance

**Comment by nikordaris** [comment]
@slavab89 and @priand I also did not have success manually calling push and remove. I agree with @slavab89 that `rerenderOnEveryChange` is probably not a good idea. The only workaround I've found is what I originally suggested plus manually injecting the field array into the component(s) that need to be updated when the array changes via `formValueSelector`.

**Comment by priand** [comment]
Thanks - I actually managed it with a simple field dealing with the array, without a FieldArray. The performance looks good so far.

**Comment by babak-bekhrad** [comment]
The flaw of this approach is that in JavaScript, **objects and arrays are reference types**, so when we get an array, we actually get a pointer to the original array's object managed by react. If we then splice it, we already mutate the original data and whilst it does work without throwing an error, this is not really how we should do it, this can lead to unpredictable apps and is definitely a bad practice. A good practice is to create a copy of the  array before manipulating it and a simple way of doing this is by calling the slice method. Slice without arguments simply copies the full array and returns a new one which is then stored. And we can now safely edit this new one and then update to react state with our new array. let me give you and example:

We have an array like this `const arr=[1,2,3,4,5]`. This is original array.

as I told you before we can do that like this:

    ```
const newVar=arr.slice();
    newVar.splice(Index,1);
    console.log(newVar);
```


**Or**

An alternative to this approach would be to use it a ES6 feature, the spread operator. You can simply set array equal to a new array and this new array can now use the spread operator which are three dots might look strange but this is a JavaScript operator. What this does is it spreads out the elements in this array into a list of elements which simply then gets added to this array, so that now we have an array, a new array, with the objects from the old array but not the old array itself.

Our prior code can be something like this: 
 

    ```
const newVar=[...arr]
    newVar.splice(Index,1);
    console.log(newVar);
```


That's it. Good luck
